### PR TITLE
Load remote stories from Tumblr API

### LINKED ISSUE #714: API for pulling stories from blog to website
Currently, our blog posts are published on Tumblr and have to be manually added to the Story model for them to appear on the website. It would be great to have an API do that for us so that we don't have outdated posts listed on the website because someone forgot to update the model after publishing a new post.

Tumblr has an [API Documentation](https://www.tumblr.com/docs/en/api/v2) that could be explored and an API developed to make this work.

### PR DESCRIPTION
**Pull request by ramonsaraiva**
Closes #714 

Draft pull request to load remote stories from Tumblr API.

~~I added a django action to load the remote stories, as we don't have a backbone for having a periodic task that would try fetching new stories and load them from time to time (I think that's the desired approach, but an action seems like a good first step).~~

image

~~The only caveat to the action is that you need to select 1 existing story to trigger it - it is possible to have actions without a queryset selected, but requires some code to change the action behavior that I felt it was a bit overkill, we can just select 1 story and trigger the action.~~

~~There's still one missing piece, for actual stories (`Your Django Story: ..` posts), that I need to load the first image in the NPF (https://www.tumblr.com/docs/npf) content. But I'll do that part once you guys agree with the approach.~~

**UPDATE:**

image

There's a management command for loading remote stories from Tumblr. I also added the functionality to download the banner image when it's a "real story" (title starts with `Your Django Story:`), cause we have thumbnails in the homepage using those images.

### CODE REVIEW

**Comment by amakarudze** [comment]
Thanks @ramonsaraiva for your work on this. 

While there is still the manual part involved in triggering the action, this is way better than actually having to enter the stories manually like we have to do now. So yes, we can use this approach for now. You can complete your work.

Sorry for late response on your PR, it was submitted during the holiday break when I had decided to unplug myself completely from the internet.

**Comment by ramonsaraiva** [comment]
No worries @amakarudze - I unplugged myself as well 😃
I'll go ahead and wrap this up in the following days then, thanks for the feedback.

**Comment by amakarudze** [comment]
> No worries @amakarudze - I unplugged myself as well 😃 I'll go ahead and wrap this up in the following days then, thanks for the feedback.

So I was trying to fix something on the server and then realised we can use commands `python manage.py ...` to do this like we do for fetching 'Your Django Story` using daily scheduled tasks on PythonAnywhere. Please see [these lines of code](https://github.com/DjangoGirls/djangogirls/blob/main/story/management/commands/fetch_stories.py#L1-L55). Sorry to provide this feedback a little too late.

**Comment by ramonsaraiva** [comment]
That works as well 👍

**Comment by amakarudze** [comment]
Hey @ramonsaraiva. Hope all is well with you. Do you have any idea on when you can move this from being a draft PR? Thanks for all your work on this.

**Comment by ramonsaraiva** [comment]
Hey @amakarudze - I'll try getting back to this week/early next week. Been busy, sorry for leaving this stale. Should be an easy wrap, just gotta move the functionality to a command, I think the only missing piece content-wise for the posts was the image. I'll move it to review once I finalize that, thanks for poking.

**Comment by ramonsaraiva** [comment]
Alrighty, I think that's how you imagined the management command to work. I'd love if you could fetch this branch locally and try it out yourself to make sure the stories are being loaded correctly.
```sh
python manage.py fetch_tumblr_stories
```
You can also run multiple times to make sure it won't re-fetch/re-create stories that already existing in our database.

If this looks good, I'll go ahead and write tests for this approach and then we can finally move to review :-)

**Comment by amakarudze** [comment]
> Alrighty, I think that's how you imagined the management command to work. I'd love if you could fetch this branch locally and try it out yourself to make sure the stories are being loaded correctly.
> 
> ```shell
> python manage.py fetch_tumblr_stories
> ```
> 
> You can also run multiple times to make sure it won't re-fetch/re-create stories that already exist in our database.
> 
> If this looks good, I'll go ahead and write tests for this approach and then we can finally move to review :-)

Hello @ramonsaraiva, I checkout this PR locally and it only pulled out only 3 stories, which are the ones I didn't already have locally so you can add tests for it. I will check the code once the tests are there. Thanks for all your work on this! 🥳✨👏

**Comment by ramonsaraiva** [comment]
@amakarudze @marksweb been ages, but finally got those tests written! Let me know if you have any suggestions or change requests. Cheers!

**Comment by marksweb** [comment]
@ramonsaraiva excellent. I saw the notification for this earlier.

I'll try to review before Djangocon Europe this week 👍

File: story/management/commands/fetch_tumblr_stories.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+from story.models import Story
+
+from core.tumblr_client import request_latest_stories
+
+
+def download_image(url: str) -> Tuple[str, File]:
+    image_name = urlparse(url).path.split("/")[-1]
+    content = urlretrieve(url)
+    return (image_name, File(open(content[0], "rb")))
```
Comment (marksweb) [inline story/management/commands/fetch_tumblr_stories.py:17]: You could drop the outer braces from this & it'll still return a `tuple`.

```suggestion
    return image_name, File(open(content[0], "rb"))
```

**Review by marksweb — CHANGES_REQUESTED** [review]
Sorry this took a bit longer than I'd hoped. I've left a few comments. Nothing significant - just things to try to keep the code nice & tidy.

File: story/services.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    )
+    to_create_urls = remote_urls.difference(in_db_urls)
+    missing_stories = [
+        remote_story
+        for remote_story in remote_stories
+        if remote_story.url in to_create_urls
+    ]
+    for missing_story in missing_stories:
+        # TODO: still missing image from story posts
+        story = Story.objects.create(
```
[the remark this answers is not in the document]
Comment (ramonsaraiva) [reply story/services.py]: I need to double check, but I'm pretty sure that if we bulk create stories, the `created` column will default to "now", unless we want to bulk create and then iterate through them to assign the proper "creation" time? We might save some writes but complexity tends to increase.

[the remark this answers is not in the document]
Comment (marksweb) [reply story/services.py]: @ramonsaraiva Yes, you're right. `bulk_create` would only help if it didn't need the additional data setting after the objects are created. I'd written this before I'd really taken that into account.

**Comment by ramonsaraiva** [comment]
@marksweb resolved the other requests, let me know if there's anything else you want me to change. Really appreciate the review.

**Comment by marksweb** [comment]
@ramonsaraiva That looks perfect now. And thank you for type hinting it all as well - just this morning I've fixed a bug on something that could easily have been avoided if it had type hints (which I've added!)

@amakarudze I think this is good to release now 👍

**Pull request merged**